config_ext_test.go

```diff
@@ -0,0 +1,461 @@
+			Populate:  &s,
+		})
+		assert.Equal(t, "", s, "unexpected result after Populate")
+	})
+
+	t.Run("scalar", func(t *testing.T) {
+		var s string
+		run(t, TestCase{
+			Value:     provider.Get("practical.toyota"),
+			HasValue:  true,
+			Interface: "camry",
+			Populate:  &s,
+		})
+		assert.Equal(t, "camry", s, "unexpected result after Populate")
+	})
+
+	t.Run("map", func(t *testing.T) {
+		var m map[string]string
+		run(t, TestCase{
+			Value:    provider.Get("practical"),
+			HasValue: true,
+			Interface: map[interface{}]interface{}{
+				"toyota": "camry",
+				"honda":  "civic",
+				"nissan": "altima",
+			},
+			Populate: &m,
+		})
+		assert.Equal(t, map[string]string{
+			"toyota": "camry",
+			"honda":  "civic",
+			"nissan": "altima",
+		}, m, "unexpected result after populate")
+	})
+
+	t.Run("slice", func(t *testing.T) {
+		var s []string
+		run(t, TestCase{
+			Value:     provider.Get("fun"),
+			HasValue:  true,
+			Interface: []interface{}{"maserati", "lamborghini"},
+			Populate:  &s,
+		})
+		assert.Equal(t, []string{"maserati", "lamborghini"}, s, "unexpected result after populate")
+	})
+
+	t.Run("struct", func(t *testing.T) {
+		type Occupants struct {
+			Driver    string
+			Passenger string
+			Backseat  []string
+		}
+		var o Occupants
+		run(t, TestCase{
+			Value:    provider.Get("occupants.honda"),
+			HasValue: true,
+			Interface: map[interface{}]interface{}{
+				"driver":    "jane",
+				"passenger": "arthur",
+				"backseat":  []interface{}{"nora"},
+			},
+			Populate: &o,
+		})
+		assert.Equal(t, Occupants{
+			Driver:    "jane",
+			Passenger: "arthur",
+			Backseat:  []string{"nora"},
+		}, o, "unexpected result after populate")
+	})
+
+	t.Run("nil merged into scalar", func(t *testing.T) {
+		// Merging replaces scalars, so an explicit nil should replace
+		// lower-priority values.
+		t.Skip("TODO: ignores explicit nil")
+		var s string
+		run(t, TestCase{
+			Value:     provider.Get("antique_scalar"),
+			HasValue:  true,
+			Interface: nil,
+			Populate:  &s,
+		})
+		assert.Empty(t, s, "unexpected result after Populate")
+	})
+
+	t.Run("nil merged into sequence", func(t *testing.T) {
+		// Merging replaces sequences, so an explicit nil should replace
+		// lower-priority values.
+		t.Skip("TODO: ignores explicit nil")
+		var s []string
+		run(t, TestCase{
+			Value:     provider.Get("antique_sequence"),
+			HasValue:  true,
+			Interface: nil,
+			Populate:  &s,
+		})
+		assert.Empty(t, s, "unexpected result after Populate")
+	})
+
+	t.Run("nil merged into mapping", func(t *testing.T) {
+		// Merging deep-merges mappings, so an explicit nil should leave
+		// lower-priority values unchanged.
+		var m map[string]string
+		run(t, TestCase{
+			Value:     provider.Get("antique_mapping"),
+			HasValue:  true,
+			Interface: map[interface{}]interface{}{"ford": "model_t"},
+			Populate:  &m,
+		})
+		assert.Equal(t, map[string]string{"ford": "model_t"}, m, "unexpected result after Populate")
+	})
```

Comment:
In Go, this is consistent with the rest of the merging behavior - maps are deep-merged, and `nil` is a valid map with no keys.

However, it's a little weird. I'm open to the idea that we treat nils and empty maps differently here - `nil` replaces the existing value, and `{}` has no effect.

Reply:
I'd prefer that behaviour, but we should probably look at what the YAML library does in this case. We may want to match it (since the rest of the library seems to try and match it)?

Reply:
Yeah - after talking through this with Alex, we came to the same conclusion. That's also how the upstream YAML lib behaves, so we're safe.